text/0000-computed-context.md

```diff
@@ -0,0 +1,329 @@
+// Total: 10
+
+stop()
+
+order.price = 3
+// Neither the computation nor autorun will be recomputed.
+```
+
+</details>
+
+## Detailed design
+
+### Technical Background
+
+Currently, `context` is how a state machine stores contextual data for its state. Context values are assigned via `assign(...)` actions, and the assignments can reference other context values:
+
+```js
+actions: assign({
+  fullName: (context) => `${context.firstName} ${context.lastName}`
+})
+```
+
+However, this is a manual and error-prone process, as it is easy to forget to do this whenever the relevant context values change. Assigning manually computed values is also not on-demand, and some of these computations can be expensive. 
+
+Another approach is to use a function:
+
+```js
+function fullName(context) {
+  return `${context.firstName} ${context.lastName}`;
+}
+
+const machine = createMachine({
+  // ...
+  on: {
+    someEvent: {
+      guard: (context) => fullName(context).trim().length > 100,
+      // ...
+    }
+  }
+});
+```
+
+This leads to important parts of the logic (computing the `fullName`, in this case) are now completely external to the machine, which can be less portable and is not easy to visualize in the future. Also, caching must be done manually.
+
+### Implementation
+
+The `computed: { ... }` property on the machine definition is a key-value pair of computed context keys to a function that takes in `state` and returns the computed context value from that state.
+
+Computed values _cannot_ be assigned to; they are read-only values.
+
+Computed values _extend_ `context` values, so their keys cannot be the same as any `context` keys.
+
+Just like getters, computed values are computed on _read_. Caching optimizations may be possible in the future: if we are able to determine which context dependencies a computed value has, we can invalidate the cached computed context value whenever any of those dependencies change on `assign(...)`.
+
+**Specifying computed context**
+
+```ts
+const machine = createMachine({
+  context: {
+    firstName: '',
+    lastName: ''
+  },
+  computed: {
+    // Computed from non-computed context
+    fullName: ({ context }) => `${context.firstName} ${context.lastName}`,
+    // Computed from computed context
+    isLongName: ({ context }) => context.fullName.length > 20,
+    // Computed from state and context
+    errorMessage: (state) => {
+      if (state.hasTag('error')) {
+        return state.context.fullName.trim().length === 0
+          ? 'Invalid name'
+          : 'Unknown error'
+      }
+
+      return null;
+    },
+    // Returns a function, e.g.:
+    // `const greeting = state.context.greet('Hello')`
+    greet: ({ context }) => (greeting: string) => `${greeting}, ${context.fullName}`
+  }
+});
+```
+
+**Reading computed context**
+
+```js
+// ...
+
+const name = state.context.fullName; // computed from context.firstName and context.lastName
+
+const greeting = state.context.greet('Hello');
```

Comment:
Shouldn't this be `state.context.greet` as a property? How is this different from `greet(state.context.firstName)`?

Reply:
I'd assume computed values will be accessed as a property, similar to accessing getters.

Reply:
This just demonstrates that a computed value can be _anything_, including a function.